### DISCUSSION TITLE
How to edit multiple projects at the same time (with rust-analyzer)

### DISCUSSION THREAD

**Discussion by setzer22** [emacs-lsp/lsp-mode#3947] [Q&A]
Hi! I'm working in Rust with `lsp-mode`, using `rust-analyzer` as the LSP server. So far I've been able to set up *almost* everything to my liking, but there's one thing VSCode does correctly out of the box and emacs doesn't. I'm not sure how I can fix it, but it's finally bugged me enough to make me dig further into this. I really don't want to go back to VSCode :grimacing:. Let me explain:

I have two Rust projects (crates) I maintain set up as two independent git repositories, and on the emacs side, these map to two separate projectile projects. And one project (let's call it `crate_a`) depends on the other (`crate_b`). During development, I like being able to develop both projects at the same time, so I temporarily set up a dependency using a `path` declaration on my `Cargo.toml`:
```toml
crate_b = { path = "../../crate_b" }
```
Doing this allows jumping to definition and finding references across the two crates, and it creates a very seamless development experience. But when I do this on emacs with `lsp-mode`, it starts two separate instances of `rust-analyzer`, one per project. This means changes I make in `crate_b` are not visible in `crate_a` unless I run `lsp-restart-workspace` (which takes a couple of minutes). For instance, if I add a new function or method in `crate_b`, the other crate won't suggest it as an auto-completion until I restart.

When calling lsp-describe session, I see two different sessions, one for each project
![image](https://user-images.githubusercontent.com/7241990/216786283-840f3474-6b2d-4f2c-847a-7fb2f744a6cf.png)

What I would like instead, is that all the folders are managed by the same LSP instance, regardless of the directory structure.

Is there a way I can achieve this?

Thanks a lot!

**Comment by yyoncho** [comment]
@flodiebold @brotzeit is rust-analyzer multi-root ready? If yes, then I guess we should simply mark it as such.

**Reply by setzer22** [reply]
I'm guessing it is, given that the same thing works on VSCode. Is there some quick test I could do to verify this on my end?

**Reply by flodiebold** [reply]
No, rust-analyzer does not support multi-root. In VSCode I'm guessing you're opening the `crate_b` workspace and then opening `crate_a` from there. That doesn't require multi-root.

**Reply by setzer22** [reply]
Oh, that's too bad then. I assumed multi-root meant being able to analyze projects from multiple roots at the same time.

> In VSCode I'm guessing you're opening the crate_b workspace and then opening crate_a from there

Yup. That's exactly what I'm doing.

So... is there a way to simulate the same behavior in emacs? I'm just looking for an easy way to tell emacs-lsp that the two folders belong to the same project / share the same project root. The only difference is that the projects are in sibling folders, and different git repos.

This is a useful feature, so it would be nice to have it, even if it's by applying some local workaround.

**Reply by yyoncho** [reply]
@setzer22 you can use lsp-rust-analyzer-library-directories . Add the second project root and it will be opened in the same server connection.

**Reply by setzer22** [reply]
Thanks @yyoncho! That seemed to do the trick! However, there are still a couple of issues to solve. I'd also be happy with workarounds for those, but it's a bit hard to me to understand where exactly I should patch stuff for this to work, so if you don't mind me asking a bit further:

- When is the `lsp-rust-analyzer-library-directories` variable initialized? How can I customize it in a way that it will work without me having to eval the code every time? For instance, if I use `use-package` with `lsp-mode` and in the `:config` section I put `(add-to-list 'lsp-rust-analyzer-library-directories "/path/to/crate_b")`, I get an error because at that point the variable is still uninitialized.
- Another issue I have, is that when something is in `lsp-rust-analyzer-library-directories`, emacs opens those files as read-only when, e.g. doing a "jump to definition" action. This works as intended for actual libraries (things in `~/.cargo/registry/src`), but I'd like to disable this behavior for the libraries I add to this list as a workaround. What do you suggest would be a good way to achieve this? :thinking: 

Thanks a lot btw! :smile:

**Reply by yyoncho** [reply]
1. Use `(with-eval-after-load 'lsp-rust ...)`
2. Let me think about it. You may do M-x view-mode to turn it off.

**Reply by setzer22** [reply]
Ok, I managed to get this working. I still appreciate it if you happen to know a less hacky way to handle this :smile: 

```lisp
;; NOTE: In order to be able to edit sibling projects seamlessly in
;; rust-analyzer, one can declare those project roots as library folders. Doing
;; this, emacs won't spawn a new rust-analyzer server for these folders, instead
;; attaching to whatever server you came from.
;;
;; This workaround was suggested to me in:
;; https://github.com/emacs-lsp/lsp-mode/discussions/3947
(defvar my-custom-rust-lsp-library-prefixes
  '("/home/user/Repositories/guee/"))

(defun my-custom-rust-register-library-folders ()
  "Registers the variables in `my-custom-rust-lsp-library-prefixes' as libraries in `lsp-rust-analyzer-library-directories'"
  (seq-doseq (lib-path my-custom-rust-lsp-library-prefixes)
    (add-to-list 'lsp-rust-analyzer-library-directories
                 lib-path)))

(use-package! lsp-rust
  :config
  (my-custom-rust-register-library-folders))

;; NOTE: When something is listed in `lsp-rust-analyzer-library-directories',
;; lsp-mode will set read-only mode when loading those files by default. We use
;; this hook to override the behavior.
(defun path-inside-p (the-path paths)
  "Returns whether `the-path` corresponds to a file or subfolder in any of the given `paths`"
  (let (result)
    (dolist (path paths result)
      (when (string-prefix-p path the-path)
        (setq result t)))))

(add-hook
 'find-file-hook
 (lambda ()
   (when (path-inside-p buffer-file-name my-custom-rust-lsp-library-prefixes)
     (read-only-mode -1))))
```

**Reply by setzer22** [reply]
> 1. Use `(with-eval-after-load 'lsp-rust ...)`

Oops, sorry @yyoncho, I had missed your message. I ended up using `use-package!` because I'm using Doom, but the two should be equivalent. What I was missing was precisely the fact that `lsp-rust` is its own package, and it's not loaded at the same time as `lsp-mode`. Thanks!

> 2. Let me think about it. You may do M-x view-mode to turn it off.

Do you think the hack I posted above could lead to issues?

**Reply by yyoncho** [reply]
The solution it should be fine. Using view-mode in lsp-mode is a bit questionable, but it may be convenient. I will leave it as it is.

**Reply by setzer22** [reply]
Nevermind... Looks like I was too hasty to mark this as solved :/

When I add my project root folder as a library path as suggested, I can jump to definition to files in `crate_b`, but only the file I navigate to seems to get loaded by emacs-lsp, so actions like `lsp-rust-find-parent-module` fail, and every time I open a different file, I am greeted by the interactive prompt of "abc.rs is not a part of any project."

This makes the workaround very limited in usability. Unless we can figure out a way to fix this new issue :thinking:

**Reply by yyoncho** [reply]
What different file means here? Is lsp-rust-find-parent-module behaving differently in vscode?

**Reply by setzer22** [reply]
Let me describe a bit better, using my example above:

- I have `crate_a`, which is the main project I start editing, and then I occasionally (mostly via "jump to definition") I end up in a file from `crate_b`.
- As suggested above, the root path for `crate_b` is in `lsp-rust-analyzer-library-directories`.
- When I open a file from `crate_b`, I get the "<filename> is not a part of any project." prompt, asking me to interactively select an option. I press `i` to select the autodetected project root.
- Then, I can edit the individual file from `crate_b`, and going back to `crate_a`, changes are shown. When checking lsp-describe-session, I can only see an lsp server, with the project root for `crate_a`. However, there's a couple of issues.
- If, from that file in `crate_b`, I try to jump to another file in `crate_b` (say, via "jump to definition"), I get the interactive prompt again. Selecting `i` every time seems to work, but it's a bit cumbersome.
- If I haven't navigated to the parent and answered the interactive prompt before, calling the function `lsp-rust-find-parent-module` returns an error. The equivalent function in VSCode has no trouble jumping to the parent folder when you're visiting a "library" file.

I think this pretty much sums it up :+1:

**Reply by setzer22** [reply]
Having thought a bit about this, I think I'd the simplest solution for me would be some kind of mode / toggle / patch that simply doesn't attempt to autodetect the current project, and only spawns a single lsp server, sending all requests to that same server. I tend to use separate emacs processes for separate projects, so the feature to have multiple LSP servers spawned by the same emacs process is not all that useful to me. I understand this is not how everyone uses emacs, but I'd be happy if you could help me find a workaround for my use case :slightly_smiling_face:
